### PR TITLE
Dismantle Mux expression / Ternary operator in dpdk

### PR DESCRIPTION
**Pull request by kamleshbhalui**
LowerExpressions pass which introduces ternary operator (?:) which is not supported in DPDK. 
This PR dismantle ternary operator to if else chain.

### CODE REVIEW

File: backends/dpdk/dpdkArch.cpp [hunk 1]
```diff
@@ -1095,6 +1095,106 @@ const IR::Node* PrependPDotToActionArgs::preorder(IR::MethodCallExpression* mce)
...
+    return tmp;
+}
+
+const IR::Expression* SimplifyTernaryOperator::addAssignment(
+    Util::SourceInfo srcInfo,
+    cstring varName,
+    const IR::Expression* expression) {
+    const IR::PathExpression* left;
+    if (auto pe = expression->to<IR::PathExpression>())
+        left = new IR::PathExpression(IR::ID(varName, pe->path->name.originalName));
```
Comment (mbudiu-vmw) [inline backends/dpdk/dpdkArch.cpp:1137]: why not the srcInfo as well?

**Review by mbudiu-vmw — CHANGES_REQUESTED** [review]

File: backends/dpdk/dpdkArch.cpp [hunk 2]
```diff
@@ -1095,6 +1095,106 @@ const IR::Node* PrependPDotToActionArgs::preorder(IR::MethodCallExpression* mce)
...
+    statements.clear();
+    visit(expression->e2);
+    auto path = addAssignment(expression->srcInfo, tmp, expression->e2);
+    auto ifFalse = statements;
+    statements = save;
+
+    auto ifStatement = new IR::IfStatement(
+        expression->e0, new IR::BlockStatement(ifTrue), new IR::BlockStatement(ifFalse));
+    statements.push_back(ifStatement);
+    typeMap->setType(path, type);
```
[the remark this answers is not in the document]
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.cpp:1118]: Yes

File: backends/dpdk/backend.cpp [hunk 3]
```diff
@@ -53,6 +53,7 @@ void DpdkBackend::convert(const IR::ToplevelBlock *tlb) {
         new P4::TypeChecking(refMap, typeMap),
         // TBD: implement dpdk lowering passes instead of reusing bmv2's lowering pass.
         new BMV2::LowerExpressions(typeMap),
+        new SimplifyTernaryOperator(typeMap, refMap),
```
[the remark this answers is not in the document]
Comment (kamleshbhalui) [reply backends/dpdk/backend.cpp]: Renamed it to DismantleMuxExpressions.
Looks like Slices taken care by LowerExpressions Pass.

File: backends/dpdk/dpdkArch.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.cpp:1118]: Yes

File: backends/dpdk/dpdkArch.cpp (hunk 1 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.cpp]: done

File: backends/dpdk/dpdkArch.cpp [hunk 4]
```diff
@@ -1095,6 +1095,106 @@ const IR::Node* PrependPDotToActionArgs::preorder(IR::MethodCallExpression* mce)
...
 
+const IR::Node* DismantleMuxExpressions::preorder(IR::Mux* expression) {
+    // We always dismantle muxes for dpdk
+    auto type = typeMap->getType(getOriginal(), true);
+    visit(expression->e0);
+    auto tmp = createTemporary(type);
+    auto save = statements;
+    statements.clear();
+    visit(expression->e1);
+    (void) addAssignment(expression->srcInfo, tmp, expression->e1);
```
Comment (mbudiu-vmw) [inline backends/dpdk/dpdkArch.cpp:1106]: Are you sure this does not need to be added to the typeMap?
What if you have an expression like a ? (b ? c : d) : (e ? f : g)

**Review by mbudiu-vmw — APPROVED** [review]

**Comment by kamleshbhalui** [comment]
@mbudiu-vmw Please merge it.

**Pull request merged**